File: voice/device.cpp
#include <iostream>
#include <cstring>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <thread>
#include <chrono>
#include <fstream>
#include <nlohmann/json.hpp>
#include <alsa/asoundlib.h>
#include <vector>
#include <ctime>

using json = nlohmann::json;
using namespace std;

// 配置参数
const char* TCP_SERVER = "192.168.1.100";
const int TCP_PORT = 3000;

// 音频参数
const int SAMPLE_RATE = 16000;
const int SAMPLE_BITS = 16;
const int CHANNELS = 1;
const int BUFFER_LENGTH = 1024;
const int VAD_THRESHOLD = 1000;

// 可靠性参数
const int MAX_RETRIES = 3;
const int HEARTBEAT_INTERVAL = 30;
const int CONNECTION_TIMEOUT = 5;
const int RECONNECT_DELAY = 5;
const char* STATE_FILE = "device_state.json";

class AudioDevice {
private:
    int tcp_socket;
    uint32_t sequence;
    time_t last_heartbeat;
    int heartbeat_failures;
    string connection_state;
    int retry_count;
    snd_pcm_t *capture_handle;
    vector<int16_t> audio_buffer;
    
    // 状态管理
    void load_state() {
        try {
            ifstream f(STATE_FILE);
            if (f.good()) {
                json state;
                f >> state;
                sequence = state["sequence"].get<uint32_t>();
            }
        } catch (const exception& e) {
            cerr << "加载状态失败: " << e.what() << endl;
        }
    }
    
    void save_state() {
        try {
            json state = {
                {"sequence", sequence},
                {"connection_state", connection_state}
            };
            ofstream f(STATE_FILE);
            f << state.dump(4);
        } catch (const exception& e) {
            cerr << "保存状态失败: " << e.what() << endl;
        }
    }
    
    bool setup_audio() {
        int err;
        if ((err = snd_pcm_open(&capture_handle, "default", SND_PCM_STREAM_CAPTURE, 0)) < 0) {
            cerr << "无法打开音频设备: " << snd_strerror(err) << endl;
            return false;
        }
        
        snd_pcm_hw_params_t *hw_params;
        snd_pcm_hw_params_alloca(&hw_params);
        snd_pcm_hw_params_any(capture_handle, hw_params);
        
        if ((err = snd_pcm_hw_params_set_access(capture_handle, hw_params, SND_PCM_ACCESS_RW_INTERLEAVED)) < 0) {
            cerr << "无法设置访问类型: " << snd_strerror(err) << endl;
            return false;
        }
        
        if ((err = snd_pcm_hw_params_set_format(capture_handle, hw_params, SND_PCM_FORMAT_S16_LE)) < 0) {
            cerr << "无法设置格式: " << snd_strerror(err) << endl;
            return false;
        }
        
        unsigned int rate = SAMPLE_RATE;
        if ((err = snd_pcm_hw_params_set_rate_near(capture_handle, hw_params, &rate, 0)) < 0) {
            cerr << "无法设置采样率: " << snd_strerror(err) << endl;
            return false;
        }
        
        if ((err = snd_pcm_hw_params_set_channels(capture_handle, hw_params, CHANNELS)) < 0) {
            cerr << "无法设置通道数: " << snd_strerror(err) << endl;
            return false;
        }
        
        if ((err = snd_pcm_hw_params(capture_handle, hw_params)) < 0) {
            cerr << "无法设置参数: " << snd_strerror(err) << endl;
            return false;
        }
        
        snd_pcm_prepare(capture_handle);
        audio_buffer.resize(BUFFER_LENGTH);
        return true;
    }
    
public:
    AudioDevice() : sequence(0), last_heartbeat(0), heartbeat_failures(0),
                   connection_state("disconnected"), retry_count(0), tcp_socket(-1) {
        load_state();
        if (!setup_audio()) {
            throw runtime_error("音频设备初始化失败");
        }
    }
    
    ~AudioDevice() {
        if (tcp_socket >= 0) {
            close(tcp_socket);
        }
        if (capture_handle) {
            snd_pcm_close(capture_handle);
        }
    }
    
    bool connect_tcp() {
        cout << "正在连接TCP服务器..." << endl;
        retry_count = 0;
        while (retry_count < MAX_RETRIES) {
            try {
                tcp_socket = socket(AF_INET, SOCK_STREAM, 0);
                if (tcp_socket < 0) {
                    throw runtime_error("创建socket失败");
                }
                
                struct timeval timeout;
                timeout.tv_sec = CONNECTION_TIMEOUT;
                timeout.tv_usec = 0;
                setsockopt(tcp_socket, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
                setsockopt(tcp_socket, SOL_SOCKET, SO_SNDTIMEO, &timeout, sizeof(timeout));
                
                struct sockaddr_in server_addr;
                server_addr.sin_family = AF_INET;
                server_addr.sin_port = htons(TCP_PORT);
                inet_pton(AF_INET, TCP_SERVER, &server_addr.sin_addr);
                
                if (connect(tcp_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
                    throw runtime_error("连接服务器失败");
                }
                
                send_connection_confirmation();
                connection_state = "connected";
                cout << "TCP服务器连接成功" << endl;
                return true;
            } catch (const exception& e) {
                cerr << "TCP连接失败: " << e.what() << endl;
                retry_count++;
                if (tcp_socket >= 0) {
                    close(tcp_socket);
                    tcp_socket = -1;
                }
                this_thread::sleep_for(chrono::seconds(RECONNECT_DELAY));
            }
        }
        return false;
    }
    
    void send_connection_confirmation() {
        uint32_t magic = 0xDEADBEEF;
        uint32_t data[3] = {htonl(magic), htonl(sequence), 0};
        if (send(tcp_socket, data, sizeof(data), 0) != sizeof(data)) {
            throw runtime_error("发送确认包失败");
        }
        
        uint32_t response;
        if (recv(tcp_socket, &response, sizeof(response), 0) != sizeof(response)) {
            throw runtime_error("服务器确认失败");
        }
    }
    
    bool detect_voice(const vector<int16_t>& audio_data) {
        try {
            // 计算短时能量
            double energy = 0;
            for (int16_t sample : audio_data) {
                energy += abs(sample);
            }
            energy /= audio_data.size();
            
            // 使用自适应阈值
            double threshold = VAD_THRESHOLD * (1.0 + get_noise_level());
            return energy > threshold;
        } catch (const exception& e) {
            cerr << "VAD检测错误: " << e.what() << endl;
            return false;
        }
    }
    
    double get_noise_level() {
        return 0.1; // 可以实现更复杂的噪声估计算法
    }
    
    bool send_audio_packet(const vector<int16_t>& audio_data) {
        retry_count = 0;
        while (retry_count < MAX_RETRIES) {
            try {
                // 构建数据包头
                uint32_t magic = 0xAABBCCDD;
                uint32_t timestamp = time(nullptr);
                uint32_t header[4] = {
                    htonl(magic),
                    htonl(sequence),
                    htonl(audio_data.size() * sizeof(int16_t)),
                    htonl(timestamp)
                };
                
                // 发送头部
                if (send(tcp_socket, header, sizeof(header), 0) != sizeof(header)) {
                    throw runtime_error("发送数据包头失败");
                }
                
                // 发送音频数据
                if (send(tcp_socket, audio_data.data(), audio_data.size() * sizeof(int16_t), 0) 
                    != audio_data.size() * sizeof(int16_t)) {
                    throw runtime_error("发送音频数据失败");
                }
                
                // 等待ACK
                uint32_t ack[2];
                if (recv(tcp_socket, ack, sizeof(ack), 0) == sizeof(ack)) {
                    uint32_t ack_seq = ntohl(ack[1]);
                    if (ack_seq == sequence) {
                        sequence++;
                        save_state();
                        return true;
                    }
                }
                
                retry_count++;
                this_thread::sleep_for(chrono::milliseconds(100));
            } catch (const exception& e) {
                cerr << "发送数据失败: " << e.what() << endl;
                retry_count++;
                this_thread::sleep_for(chrono::milliseconds(100));
            }
        }
        return false;
    }
    
    bool send_heartbeat() {
        try {
            time_t current_time = time(nullptr);
            if (current_time - last_heartbeat >= HEARTBEAT_INTERVAL) {
                uint32_t magic = 0xFFEEDDCC;
                uint32_t data[3] = {
                    htonl(magic),
                    htonl(sequence),
                    htonl(static_cast<uint32_t>(current_time))
                };
                
                if (send(tcp_socket, data, sizeof(data), 0) != sizeof(data)) {
                    throw runtime_error("发送心跳包失败");
                }
                
                uint32_t response[2];
                if (recv(tcp_socket, response, sizeof(response), 0) == sizeof(response) &&
                    ntohl(response[0]) == magic) {
                    last_heartbeat = current_time;
                    heartbeat_failures = 0;
                    return true;
                } else {
                    heartbeat_failures++;
                }
                
                if (heartbeat_failures >= MAX_RETRIES) {
                    throw runtime_error("心跳检测失败次数过多");
                }
            }
            return true;
        } catch (const exception& e) {
            cerr << "发送心跳包失败: " << e.what() << endl;
            heartbeat_failures++;
            return false;
        }
    }
    
    void run() {
        while (true) {
            try {
                // 读取音频数据
                int err = snd_pcm_readi(capture_handle, audio_buffer.data(), BUFFER_LENGTH);
                if (err == -EPIPE) {
                    snd_pcm_prepare(capture_handle);
                    continue;
                } else if (err < 0) {
                    throw runtime_error("读取音频数据失败: " + string(snd_strerror(err)));
                }
                
                // VAD检测
                if (detect_voice(audio_buffer)) {
                    if (!send_audio_packet(audio_buffer)) {
                        throw runtime_error("发送音频数据失败");
                    }
                }
                
                // 心跳检测
                if (!send_heartbeat()) {
                    throw runtime_error("心跳检测失败");
                }
                
            } catch (const exception& e) {
                cerr << "运行错误: " << e.what() << endl;
                connection_state = "disconnected";
                save_state();
                reconnect();
            }
        }
    }
    
    void reconnect() {
        cout << "正在尝试重新连接..." << endl;
        if (tcp_socket >= 0) {
            close(tcp_socket);
            tcp_socket = -1;
        }
        
        while (true) {
            if (connect_tcp()) {
                connection_state = "connected";
                save_state();
                break;
            }
            this_thread::sleep_for(chrono::seconds(RECONNECT_DELAY));
        }
        cout << "重新连接成功" << endl;
    }
};

int main() {
    try {
        AudioDevice device;
        
        // 初始化连接
        while (true) {
            try {
                if (device.connect_tcp()) {
                    break;
                }
            } catch (const exception& e) {
                cerr << "初始化失败: " << e.what() << endl;
                this_thread::sleep_for(chrono::seconds(RECONNECT_DELAY));
            }
        }
        
        // 运行主循环
        device.run();
        
    } catch (const exception& e) {
        cerr << "程序错误: " << e.what() << endl;
        return 1;
    }
    
    return 0;
} 

File: voice/device.py
import machine
import network
import socket
import time
import json
from machine import I2S, Pin
import struct
import os

# 配置参数
WIFI_SSID = "your_wifi_ssid"
WIFI_PASSWORD = "your_wifi_password"
TCP_SERVER = "192.168.1.100"
TCP_PORT = 3000

# 音频参数
SAMPLE_RATE = 16000
SAMPLE_BITS = 16
CHANNELS = 1
BUFFER_LENGTH = 1024
VAD_THRESHOLD = 1000  # VAD能量阈值

# 可靠性参数
MAX_RETRIES = 3
HEARTBEAT_INTERVAL = 30
CONNECTION_TIMEOUT = 5
RECONNECT_DELAY = 5
STATE_FILE = 'device_state.json'

class AudioDevice:
    def __init__(self):
        self.wifi = network.WLAN(network.STA_IF)
        self.tcp_socket = None
        self.sequence = 0
        self.last_heartbeat = 0
        self.heartbeat_failures = 0
        self.connection_state = 'disconnected'
        self.retry_count = 0
        
        # 加载保存的状态
        self.load_state()
        
        # 配置I2S
        self.i2s = I2S(
            I2S.NUM0,
            bck=Pin(26),
            ws=Pin(25),
            sdin=Pin(27),
            standard=I2S.PHILIPS,
            mode=I2S.MASTER_RX,
            dataformat=I2S.B16,
            channelformat=I2S.MONO,
            samplerate=SAMPLE_RATE,
            dmacount=8,
            dmalen=BUFFER_LENGTH
        )
        
        # 初始化音频缓冲区
        self.audio_buffer = bytearray(BUFFER_LENGTH * 2)
        
    def load_state(self):
        try:
            if STATE_FILE in os.listdir():
                with open(STATE_FILE, 'r') as f:
                    state = json.load(f)
                    self.sequence = state.get('sequence', 0)
        except:
            pass
            
    def save_state(self):
        try:
            state = {
                'sequence': self.sequence,
                'connection_state': self.connection_state
            }
            with open(STATE_FILE, 'w') as f:
                json.dump(state, f)
        except:
            pass
        
    def connect_wifi(self):
        print("正在连接WiFi...")
        if not self.wifi.active():
            self.wifi.active(True)
        
        retry_count = 0
        while not self.wifi.isconnected() and retry_count < MAX_RETRIES:
            try:
                self.wifi.connect(WIFI_SSID, WIFI_PASSWORD)
                start_time = time.time()
                while not self.wifi.isconnected():
                    if time.time() - start_time > CONNECTION_TIMEOUT:
                        raise Exception("WiFi连接超时")
                    time.sleep(1)
                print(f"WiFi已连接，IP: {self.wifi.ifconfig()[0]}")
                return True
            except Exception as e:
                print(f"WiFi连接失败: {e}")
                retry_count += 1
                time.sleep(RECONNECT_DELAY)
        
        return self.wifi.isconnected()
        
    def connect_tcp(self):
        print("正在连接TCP服务器...")
        retry_count = 0
        while retry_count < MAX_RETRIES:
            try:
                self.tcp_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                self.tcp_socket.settimeout(CONNECTION_TIMEOUT)
                self.tcp_socket.connect((TCP_SERVER, TCP_PORT))
                
                # 发送连接确认包
                self.send_connection_confirmation()
                self.connection_state = 'connected'
                print("TCP服务器连接成功")
                return True
            except Exception as e:
                print(f"TCP连接失败: {e}")
                retry_count += 1
                if self.tcp_socket:
                    self.tcp_socket.close()
                time.sleep(RECONNECT_DELAY)
        return False
    
    def send_connection_confirmation(self):
        confirmation = struct.pack(">III", 0xDEADBEEF, self.sequence, 0)
        self.tcp_socket.send(confirmation)
        # 等待服务器确认
        response = self.tcp_socket.recv(1024)
        if not response or len(response) < 4:
            raise Exception("服务器确认失败")
            
    def detect_voice(self, audio_data):
        # 增强的VAD检测
        try:
            # 计算短时能量
            energy = sum(abs(x) for x in struct.unpack(f"<{len(audio_data)//2}h", audio_data))
            energy /= (len(audio_data) // 2)
            
            # 使用自适应阈值
            threshold = VAD_THRESHOLD * (1.0 + self.get_noise_level())
            return energy > threshold
        except Exception as e:
            print(f"VAD检测错误: {e}")
            return False
            
    def get_noise_level(self):
        # 简单的背景噪声估计
        return 0.1  # 可以实现更复杂的噪声估计算法
        
    def send_audio_packet(self, audio_data):
        retry_count = 0
        while retry_count < MAX_RETRIES:
            try:
                # 构建数据包
                header = struct.pack(">IIII", 0xAABBCCDD, self.sequence, len(audio_data), int(time.time()))
                packet = header + audio_data
                self.tcp_socket.send(packet)
                
                # 等待ACK
                ack = self.tcp_socket.recv(8)
                if len(ack) == 8:
                    ack_seq = struct.unpack(">II", ack)[1]
                    if ack_seq == self.sequence:
                        self.sequence += 1
                        self.save_state()  # 保存状态
                        return True
                
                retry_count += 1
                time.sleep(0.1)
            except Exception as e:
                print(f"发送数据失败: {e}")
                retry_count += 1
                time.sleep(0.1)
        return False
        
    def send_heartbeat(self):
        try:
            current_time = time.time()
            if current_time - self.last_heartbeat >= HEARTBEAT_INTERVAL:
                heartbeat = struct.pack(">III", 0xFFEEDDCC, self.sequence, int(current_time))
                self.tcp_socket.send(heartbeat)
                
                # 等待心跳确认
                response = self.tcp_socket.recv(8)
                if len(response) == 8 and struct.unpack(">II", response)[0] == 0xFFEEDDCC:
                    self.last_heartbeat = current_time
                    self.heartbeat_failures = 0
                    return True
                else:
                    self.heartbeat_failures += 1
                    
                if self.heartbeat_failures >= MAX_RETRIES:
                    raise Exception("心跳检测失败次数过多")
                    
            return True
        except Exception as e:
            print(f"发送心跳包失败: {e}")
            self.heartbeat_failures += 1
            return False
            
    def run(self):
        while True:
            try:
                # 读取音频数据
                num_bytes = self.i2s.readinto(self.audio_buffer)
                if num_bytes > 0:
                    # VAD检测
                    if self.detect_voice(self.audio_buffer[:num_bytes]):
                        if not self.send_audio_packet(self.audio_buffer[:num_bytes]):
                            raise Exception("发送音频数据失败")
                
                # 心跳检测
                if not self.send_heartbeat():
                    raise Exception("心跳检测失败")
                    
            except Exception as e:
                print(f"运行错误: {e}")
                self.connection_state = 'disconnected'
                self.save_state()
                self.reconnect()
                
    def reconnect(self):
        print("正在尝试重新连接...")
        if self.tcp_socket:
            self.tcp_socket.close()
        
        while True:
            if not self.wifi.isconnected():
                if not self.connect_wifi():
                    time.sleep(RECONNECT_DELAY)
                    continue
                    
            if self.connect_tcp():
                self.connection_state = 'connected'
                self.save_state()
                break
                
            time.sleep(RECONNECT_DELAY)
        print("重新连接成功")

def main():
    device = AudioDevice()
    
    # 初始化连接
    while True:
        try:
            if device.connect_wifi() and device.connect_tcp():
                break
        except Exception as e:
            print(f"初始化失败: {e}")
            time.sleep(RECONNECT_DELAY)
    
    # 运行主循环
    device.run()

if __name__ == "__main__":
    main() 